pkg/google/networking/pricing_map.go

```diff
@@ -0,0 +1,190 @@
+	ErrUnknownDescription              = errors.New("unknown description")
+)
+
+type ParsedSkuData struct {
+	Region      string
+	Price       float64
+	Description string
+}
+
+func NewParsedSkuData(region string, price float64, description string) *ParsedSkuData {
+	return &ParsedSkuData{
+		Region:      region,
+		Price:       price,
+		Description: description,
+	}
+}
+
+type pricing struct {
+	forwardingRuleCost        float64
+	inboundDataProcessedCost  float64
+	outboundDataProcessedCost float64
+}
+
+func NewPricing() *pricing {
+	return &pricing{
+		forwardingRuleCost:        0,
+		inboundDataProcessedCost:  0,
+		outboundDataProcessedCost: 0,
+	}
+}
+
+type pricingMap struct {
+	pricing   map[string]*pricing
+	logger    *slog.Logger
+	gcpClient client.Client
+}
+
+func newPricingMap(logger *slog.Logger, gcpClient client.Client) (*pricingMap, error) {
+	pm := &pricingMap{
+		pricing:   make(map[string]*pricing),
+		logger:    logger,
+		gcpClient: gcpClient,
+	}
+
+	if err := pm.populate(context.Background()); err != nil {
+		return nil, err
+	}
+	return pm, nil
+}
+
+func (pm *pricingMap) getServiceName(ctx context.Context) (string, error) {
+	serviceName, err := pm.gcpClient.GetServiceName(ctx, serviceName)
+	if err != nil {
+		return "", fmt.Errorf("%w: %w", ErrGettingNetworkingServiceName, err)
+	}
+	return serviceName, nil
+}
+
+func (pm *pricingMap) getPricing(ctx context.Context) ([]*billingpb.Sku, error) {
+	serviceName, err := pm.getServiceName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	skus := pm.gcpClient.GetPricing(ctx, serviceName)
+	if len(skus) == 0 {
+		return nil, fmt.Errorf("%w: %w", ErrNoSKUsFoundForNetworkingService, err)
+	}
+	return skus, nil
+}
+
+func (pm *pricingMap) populate(ctx context.Context) error {
+	pm.logger.LogAttrs(ctx, slog.LevelInfo, "Populating pricing map")
+	skus, err := pm.getPricing(ctx)
+	if err != nil {
+		return err
+	}
+	skuData, err := pm.parseSku(skus)
+	if err != nil {
+		return err
+	}
+	if err := pm.processSkuData(skuData); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (pm *pricingMap) parseSku(skus []*billingpb.Sku) ([]*ParsedSkuData, error) {
+	var skuData []*ParsedSkuData
+	for _, sku := range skus {
+		if sku.Category.ResourceGroup == ResourceGroup {
+			if len(sku.GeoTaxonomy.Regions) == 0 || len(sku.PricingInfo) == 0 || len(sku.PricingInfo[0].PricingExpression.TieredRates) == 0 {
+				continue
+			}
+			region := sku.GeoTaxonomy.Regions[0]
+			price := float64(sku.PricingInfo[0].PricingExpression.TieredRates[0].UnitPrice.Nanos) / 1e9
+
+			for _, desc := range []string{
+				forwardingRuleDescription,
+				outboundDataProcessedDescription,
+				inboundDataProcessedDescription,
+			} {
+				if strings.Contains(sku.Description, desc) {
+					skuData = append(skuData, NewParsedSkuData(region, price, desc))
+				}
+			}
+		}
+	}
+	return skuData, nil
+}
+
+func (pm *pricingMap) processSkuData(skuData []*ParsedSkuData) error {
```

Comment:
protecting writes to the pricingMap with a lock
```suggestion
func (pm *pricingMap) processSkuData(skuData []*ParsedSkuData) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()
```